pkg/infra/benchmark_test.go

```diff
@@ -0,0 +1,77 @@
+package infra
+
+import (
+	"net"
+	"testing"
+
+	"github.com/guoger/tape/e2e/mock"
+	"github.com/hyperledger/fabric-protos-go/peer"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+func StartMockPeer() (*mock.Server, string) {
+	lis, _ := net.Listen("tcp", "127.0.0.1:0")
+	grpcServer := grpc.NewServer()
+	mockPeer := &mock.Server{GrpcServer: grpcServer, Listener: lis}
+	go mockPeer.Start()
+	return mockPeer, lis.Addr().String()
+}
+
+func StartProposer(signed, processed chan *Elements, done chan struct{}, logger *log.Logger, threshold int, addr string) {
+	peer := Node{
+		Addr: addr,
+	}
+	Proposer, _ := CreateProposer(peer, logger)
+	go Proposer.Start(signed, processed, done, threshold)
+}
+
+func GetData() *Elements {
+	return &Elements{
+		SignedProp: &peer.SignedProposal{},
+	}
+}
+
+// suggestion mainter checkout pr and test locally to verify the preformance benchmark.
+// for any preformance impacts features.
+// go test --bench=. -v ./... --count=1
+
+// preformance history, just logging previous time running the benchmark below at local.
+// 2020-Nov | Sam | 113753 ns/op	    9226 B/op	     184 allocs/op
+func benchmarkNPeer(concurrent int, b *testing.B) {
+	processed := make(chan *Elements, 10)
+	done := make(chan struct{})
+	signeds := make([]chan *Elements, concurrent)
+	for i := 0; i < concurrent; i++ {
+		signeds[i] = make(chan *Elements, 10)
+		mockpeer, mockpeeraddr := StartMockPeer()
+		StartProposer(signeds[i], processed, done, nil, concurrent, mockpeeraddr)
+		defer mockpeer.Stop()
+	}
+
+	b.ResetTimer()
+	go func() {
+		for i := 0; i < b.N; i++ {
+			data := &Elements{SignedProp: &peer.SignedProposal{}}
+			for _, s := range signeds {
+				s <- data
+			}
+		}
+	}()
+	var n int
+	for n < b.N {
+		<-processed
+		n++
+	}
+	b.StopTimer()
+	done <- struct{}{}
```

Comment:
you need to `close` channel since there are multiple listeners. And `defer` is always a good practice